plugins/completion/nvim-cmp/cmp-helpers.nix

```diff
@@ -1,54 +1,119 @@
-    "vim_lsp" = "cmp-vim-lsp";
-    "path" = "cmp-path";
-    "git" = "cmp-git";
-    "conventionalcommits" = "cmp-conventionalcommits";
     "cmdline" = "cmp-cmdline";
-    "cmp-cmdline-history" = "cmp-cmdline-history";
-    "fuzzy_buffer" = "cmp-fuzzy-buffer";
-    "fuzzy_path" = "cmp-fuzzy-path";
-    "rg" = "cmp-rg";
-    "fish" = "cmp-fish";
-    "tmux" = "cmp-tmux";
-    "zsh" = "cmp-zsh";
-    "crates" = "crates-nvim";
-    "npm" = "cmp-npm";
     "cmp-clippy" = "cmp-clippy";
+    "cmp-cmdline-history" = "cmp-cmdline-history";
+    "cmp_pandoc" = "cmp-pandoc-nvim";
     "cmp_tabnine" = "cmp-tabnine";
+    "conventionalcommits" = "cmp-conventionalcommits";
     "copilot" = "cmp-copilot";
+    "crates" = { package = "crates-nvim"; extraConfig = "require('crates').setup()"; };
     "dap" = "cmp-dap";
+    "dictionary" = "cmp-dictionary";
+    "digraphs" = "cmp-digraphs";
     "emoji" = "cmp-emoji";
+    "fish" = "cmp-fish";
+    "fuzzy_buffer" = "cmp-fuzzy-buffer";
+    "fuzzy_path" = "cmp-fuzzy-path";
+    "git" = "cmp-git";
     "greek" = "cmp-greek";
     "latex_symbols" = "cmp-latex-symbols";
     "look" = "cmp-look";
+    "luasnip" = "cmp-luasnip";
+    "npm" = "cmp-npm";
+    "nvim_lsp" = "cmp-nvim-lsp";
+    "nvim_lsp_document_symbol" = "cmp-nvim-lsp-document-symbol";
+    "nvim_lsp_signature_help" = "cmp-nvim-lsp-signature-help";
     "nvim_lua" = "cmp-nvim-lua";
+    "omni" = "cmp-omni";
     "pandoc_references" = "cmp-pandoc-references";
-    "cmp_pandoc" = "cmp-pandoc-nvim";
+    "path" = "cmp-path";
+    "rg" = "cmp-rg";
+    "snippy" = "cmp-snippy";
+    "spell" = "cmp-spell";
+    "tmux" = "cmp-tmux";
     "treesitter" = "cmp-treesitter";
+    "ultisnips" = "cmp-nvim-ultisnips";
+    "vim_lsp" = "cmp-vim-lsp";
     "vimwiki-tags" = "cmp-vimwiki-tags";
+    "vsnip" = "cmp-vsnip";
+    "zsh" = "cmp-zsh";
   };
+
+  # create a list off all packages of all activated sources for extraPackgs
+  sourcePackages = attrs:
+    let
+      packageList = mapAttrsToList (sourceName: _option:
+        if isNull attrs.${sourceName} then
+          null
+        else if attrs.${sourceName}.enable then
+          let
+            value = sourceNameAndPlugin.${sourceName};
+            package =
+              if isString value then
+                value
+              else
+                value.package;
+          in pkgs.vimExtraPlugins.${package}
+        else
+          null
+        ) attrs;
+    in removeNull packageList;
+
+  # create a list off all activated sources for cmp lua config
+  sourceConfig = attrs:
+    let
+      packageList = mapAttrsToList (sourceName: _option:
+        if isNull attrs.${sourceName} then
+          null
+        else if attrs.${sourceName}.enable then
+          { name = sourceName; }
+        else
+          null
+        ) attrs;
+    in removeNull packageList;
+
+  # create a list off all extraConfig of activated sources for extraConfig of cmp module
+  sourceExtraConfig = attrs:
```

Comment:
Could you please explain what these changes are for? Basically, what was wrong with what we had before and why do you feel the need for these.

Reply:
I am not sure to which lines exactly you are referring.
I added these three helper functions to generate lists for all sources that are activated.
If I understand your version, the user adds the sources like in the lua config and your code automatically adds the correct packages.

In my version the sources are all options which can be enabled in nix. These three helper functions help to add the correct packges, the correct config to lua, and extraConfig if needed.

Reply:
Yes, that's what I was asking, thanks for clarifying! I think your solution is better, will need to find a way to keep it backwards-compatible though, really want to avoid breaking configs

Reply:
As far as I can see, these are only internal changes. They should not be able to break any config.
The breaking changes are in `cmp/default.nix`/`sources.nix`
I will try to make it backwards compatible